src/test/java/com/amazon/ion/impl/macro/MacroEvaluatorTest.kt

```diff
@@ -0,0 +1,582 @@
+                StringValue(emptyList(), "b"),
+                StringValue(emptyList(), "c"),
+            )
+        )
+
+        assertEquals(StringValue(emptyList(), "abc"), evaluator.expandNext())
+        assertEquals(null, evaluator.expandNext())
+    }
+
+    @Test
+    fun `nested make_string`() {
+        // Given: <system macros>
+        // When:
+        //   (:make_string "a" (:make_string "b" "c" "d"))
+        // Then:
+        //   "abcd"
+
+        val evaluator = evaluatorWithMacroTable(Macros.MAKE_STRING)
+
+        evaluator.initExpansion(
+            listOf(
+                EExpression(MacroRef.ByName("make_string"), 0, 8),
+                ExpressionGroup(1, 8),
+                StringValue(emptyList(), "a"),
+                EExpression(MacroRef.ByName("make_string"), 3, 8),
+                ExpressionGroup(4, 8),
+                StringValue(emptyList(), "b"),
+                StringValue(emptyList(), "c"),
+                StringValue(emptyList(), "d"),
+            )
+        )
+
+        assertEquals(StringValue(emptyList(), "abcd"), evaluator.expandNext())
+        assertEquals(null, evaluator.expandNext())
+    }
+
+    @Test
+    fun `macro with a variable substitution in struct field position`() {
+        // Given:
+        //   (macro foo_struct (x*) {foo: x})
+        // When:
+        //   (:foo_struct bar)
+        // Then:
+        //   {foo: bar}
+
+        val evaluator = evaluatorWithMacroTable(Macros.FOO_STRUCT)
+
+        evaluator.initExpansion(
+            listOf(
+                EExpression(MacroRef.ByName("foo_struct"), 0, 2),
+                StringValue(value = "bar")
+            )
+        )
+
+        assertIsInstance<StructValue>(evaluator.expandNext())
+        evaluator.stepIn()
+        assertEquals(FieldName(FakeSymbolToken("foo", -1)), evaluator.expandNext())
+        assertEquals(StringValue(value = "bar"), evaluator.expandNext())
+        assertEquals(null, evaluator.expandNext())
+        evaluator.stepOut()
+        assertEquals(null, evaluator.expandNext())
+    }
+
+    @Test
+    fun `macro with a variable substitution in struct field position with multiple arguments`() {
+        // Given:
+        //   (macro foo_struct (x*) {foo: x})
+        // When:
+        //   (:foo_struct (: bar baz))
+        // Then:
+        //   {foo: bar, foo: baz}
+
+        val evaluator = evaluatorWithMacroTable(Macros.FOO_STRUCT)
+
+        evaluator.initExpansion(
+            listOf(
+                EExpression(MacroRef.ByName("foo_struct"), 0, 4),
+                ExpressionGroup(1, 4),
+                StringValue(value = "bar"),
+                StringValue(value = "baz")
+            )
+        )
+
+        assertIsInstance<StructValue>(evaluator.expandNext())
+        evaluator.stepIn()
+        // Yes, the field name should be here only once. The Ion reader that wraps the evaluator
+        // is responsible for carrying the field name over to any values that follow.
```

Comment:
That's the API that I came up with. 🤷‍♂️ 

To make it easier for me to ensure correctness, I started by having _very_ segregated concerns. The `MacroEvaluator` class is only concerned with combining `Expression`s. The `MacroEvaluatorAsIonReader` class is responsible for taking the `MacroEvaluator` output and making it fit the `IonReader` expectations. It doesn't have to be this way, and if we merge the `MacroEvaluatorAsIonReader` class with `MacroEvaluator`, then this will no longer be relevant.